service/storage_service.cc

```diff
@@ -830,6 +830,84 @@ void storage_service::scan_cdc_generations() {
     }
 }
 
+future<> storage_service::check_and_repair_cdc_streams() {
+    return async([this] { 
+        auto latest = _cdc_streams_ts;
+        const auto& endpoint_states = _gossiper.get_endpoint_states();
+        for (const auto& [addr, state] : endpoint_states) {
+            if (!_gossiper.is_normal(addr))  {
+                throw std::runtime_error(format("All nodes must be in NORMAL state while performing check_and_repair_cdc_streams"
+                        " ({} is in state {})", addr, _gossiper.get_gossip_status(state)));
+            }
+
+            const auto ts = cdc::get_streams_timestamp_for(addr, _gossiper);
+            if (!latest || (ts && *ts > *latest)) {
+                latest = ts;
+            }
+        }
+
+        bool should_regenerate = false;
+        std::optional<cdc::topology_description> gen;
+        try {
+            gen = _sys_dist_ks.local().read_cdc_topology_description(
+                    *latest, { _token_metadata.count_normal_token_owners() }).get0();
+        } catch (exceptions::request_timeout_exception& e) {
+            throw; // Stop and inform the API caller about timeout
+        } catch (exceptions::unavailable_exception& e) {
+            throw; // Stop and inform the API caller about unavailability
```

Comment:
I checked how it works in practice:
```
nodetool: Scylla API server HTTP POST to URL '/storage_service/cdc_streams_check_and_repair' failed: exceptions::unavailable_exception (Cannot achieve consistency level for cl QUORUM. Requires 2, alive 1)
See 'nodetool help' or 'nodetool help <command>'.
```

I'm afraid it could leave the user very confused. What happened? Why didn't it work? I didn't do any QUORUM requests, why is it mentioning QUORUM?

We should wrap the error, adding a comment for the user, with some guidance what to do next.

For example, we could tell him that the request failed because either some nodes are down or unreachable due to a networking issue, and that he should fix the network/restart the node and retry the command. If a node is unrecoverable, the user should first remove the node from the cluster and then retry the command.

The specific internal error should also be printed, but it's information for field engineers/developers more than for the user himself.

(this review comment applies to the timeout cases too).

Reply:
Please have a look now. I throw `request_execution_exception` at the user, which arrives at nodetool with something like this:
```
nodetool: Scylla API server HTTP POST to URL '/storage_service/cdc_streams_check_and_repair' failed: exceptions::request_execution_exception (Timeout while fetching CDC topology description. Note: this is likely caused by node(s) being down or unreachable. It is recommended to check the network and restart/remove the failed node(s), then retry checkAndRepairCdcStreams command.)
```
...while the `cdc_log` gets such a message:
```
ERROR 2020-06-06 16:47:39,468 [shard 0] cdc - Timeout while fetching CDC topology description: "Operation timed out for ks.cf - received only 666 responses from 777 CL=TWO.". Translating the exception to `request_execution_exception`.
```